homeassistant/components/google_photos/services.py

```diff
@@ -0,0 +1,116 @@
+
+__all__ = [
+    "DOMAIN",
+]
+
+CONF_CONFIG_ENTRY_ID = "config_entry_id"
+
+UPLOAD_SERVICE = "upload"
+UPLOAD_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_CONFIG_ENTRY_ID): cv.string,
+        vol.Required(CONF_FILENAME): vol.All(cv.ensure_list, [cv.string]),
+    }
+)
+
+
+def _read_file_contents(
+    hass: HomeAssistant, filenames: list[str]
+) -> list[tuple[str, bytes]]:
+    """Read the mime type and contents from each filen."""
+    results = []
+    for filename in filenames:
+        if not hass.config.is_allowed_path(filename):
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="no_access_to_path",
+                translation_placeholders={"filename": filename},
+            )
+        filename_path = Path(filename)
+        if not filename_path.exists():
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="filename_does_not_exist",
+                translation_placeholders={"filename": filename},
+            )
+        mime_type, _ = mimetypes.guess_type(filename)
+        if mime_type is None or not (mime_type.startswith(("image", "video"))):
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="filename_is_not_image",
+                translation_placeholders={"filename": filename},
+            )
+        results.append((mime_type, filename_path.read_bytes()))
+    return results
+
+
+def async_register_services(hass: HomeAssistant) -> None:
+    """Register Google Photos services."""
+
+    async def async_handle_upload(call: ServiceCall) -> ServiceResponse:
+        """Generate content from text and optionally images."""
+        config_entry: GooglePhotosConfigEntry | None = (
+            hass.config_entries.async_get_entry(call.data[CONF_CONFIG_ENTRY_ID])
+        )
+        if not config_entry:
+            raise ServiceValidationError(
+                translation_domain=DOMAIN,
+                translation_key="integration_not_found",
+                translation_placeholders={"target": DOMAIN},
+            )
+        scopes = config_entry.data["token"]["scope"].split(" ")
+        if UPLOAD_SCOPE not in scopes:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="missing_upload_permission",
+                translation_placeholders={"target": DOMAIN},
+            )
+
+        client_api = config_entry.runtime_data
+        upload_tasks = []
+        file_results = await hass.async_add_executor_job(
+            _read_file_contents, hass, call.data[CONF_FILENAME]
+        )
+        for mime_type, content in file_results:
+            upload_tasks.append(client_api.upload_content(content, mime_type))
+        upload_tokens = await asyncio.gather(*upload_tasks)
+        media_ids = await client_api.create_media_items(upload_tokens)
```

Comment:
Do you want to support partial successes? Right now if a single upload fails the code will never call batchCreate. You could call batchCreate for the ones that succeeded but then you would also need to change the API response at line 105 below. Not sure it's worth the complexity.

Reply:
I don't want to support partial success given it's an advanced technique for callers to handle. While it would be more efficient, it would mean the API for service responses is more complex and was decided against generally in https://github.com/home-assistant/architecture/discussions/777

Maybe could be considered if some sort of on_error handler is implemented
https://github.com/home-assistant/architecture/discussions/845 that way the default behavior is fail closed with an option to resume if needed?

An alternative could be to not support multiple files at all in a single call which would also be ok...